dali/ft/filecopy.cpp

```diff
@@ -465,29 +465,34 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
 
 bool FileSizeThread::wait(unsigned timems)
 {
-    while (!sem.wait(timems))   { // report every time
+    while (!sem.wait(timems))
+    { // report every time
         if (!cur.get())
             continue;       // window here?
         cur->Link();
         RemoteFilename *rfn=NULL;
-        if (copy) {
+        if (copy)
+        {
             if (!cur->mirrorFilename.isNull())
                 rfn = &cur->mirrorFilename;
         }
-        else {
+        else
+        {
             rfn = &cur->filename;
         }
-        if (rfn) {
+        if (rfn)
+        {
             StringBuffer url;
             WARNLOG("Waiting for file: %s",rfn->getRemotePath(url).str());
-            cur->Release();
+            if (cur.get())
```

Comment:
This looks like the change you made, but if cur is null it will have crashed earlier on the cur->link() I would have thought (but then again, if cur is null then the whole loop should have been skipped by the continue).

If some other thread is modifying cur concurrently, then this code has bigger problems than you can solve with a simple check for non-null.

Reply:
Agree, it shouldn't be here if cur was null, so this additional check should not be needed/helpful.
The fact it is null (backtrace suggests it is) at this point, suggests there's a race condition or overlapping thread, that's clearing cur, after the if !cur.get() test.

Reply:
I saw in my attempts to reproduce this problem the value of cur can changed by FileSizeThread::run() during the WARNLOG(...) finished.

Reply:
This code is essentially not thread safe afaics.
At any point during FileSizeThread::wait, cur could be cleared, so the crash seen was one of a number of possible crashes.
The fix should ensure/protect the object FileSizeThread::wait is dealing with cannot be altered, e.g. by using a CS lock whilst it a link copy of cur.